src/avx512-64bit-keyvaluesort.hpp

```diff
@@ -0,0 +1,909 @@
+    zmm_vector<uint64_t>::storeu(indexes + 32, index_zmm[4]);
+    zmm_vector<uint64_t>::storeu(indexes + 40, index_zmm[5]);
+    zmm_vector<uint64_t>::storeu(indexes + 48, index_zmm[6]);
+    zmm_vector<uint64_t>::storeu(indexes + 56, index_zmm[7]);
+    zmm_vector<uint64_t>::mask_storeu(indexes + 64, load_mask1, index_zmm[8]);
+    zmm_vector<uint64_t>::mask_storeu(indexes + 72, load_mask2, index_zmm[9]);
+    zmm_vector<uint64_t>::mask_storeu(indexes + 80, load_mask3, index_zmm[10]);
+    zmm_vector<uint64_t>::mask_storeu(indexes + 88, load_mask4, index_zmm[11]);
+    zmm_vector<uint64_t>::mask_storeu(indexes + 96, load_mask5, index_zmm[12]);
+    zmm_vector<uint64_t>::mask_storeu(indexes + 104, load_mask6, index_zmm[13]);
+    zmm_vector<uint64_t>::mask_storeu(indexes + 112, load_mask7, index_zmm[14]);
+    zmm_vector<uint64_t>::mask_storeu(indexes + 120, load_mask8, index_zmm[15]);
+
+    vtype::storeu(keys, key_zmm[0]);
+    vtype::storeu(keys + 8, key_zmm[1]);
+    vtype::storeu(keys + 16, key_zmm[2]);
+    vtype::storeu(keys + 24, key_zmm[3]);
+    vtype::storeu(keys + 32, key_zmm[4]);
+    vtype::storeu(keys + 40, key_zmm[5]);
+    vtype::storeu(keys + 48, key_zmm[6]);
+    vtype::storeu(keys + 56, key_zmm[7]);
+    vtype::mask_storeu(keys + 64, load_mask1, key_zmm[8]);
+    vtype::mask_storeu(keys + 72, load_mask2, key_zmm[9]);
+    vtype::mask_storeu(keys + 80, load_mask3, key_zmm[10]);
+    vtype::mask_storeu(keys + 88, load_mask4, key_zmm[11]);
+    vtype::mask_storeu(keys + 96, load_mask5, key_zmm[12]);
+    vtype::mask_storeu(keys + 104, load_mask6, key_zmm[13]);
+    vtype::mask_storeu(keys + 112, load_mask7, key_zmm[14]);
+    vtype::mask_storeu(keys + 120, load_mask8, key_zmm[15]);
+}
+
+template <typename vtype, typename type_t>
+X86_SIMD_SORT_INLINE type_t get_pivot_64bit(type_t *keys,
+                                            uint64_t *indexes,
+                                            const int64_t left,
+                                            const int64_t right)
+{
+    // median of 8
+    int64_t size = (right - left) / 8;
+    using zmm_t = typename vtype::zmm_t;
+    using index_type = zmm_vector<uint64_t>::zmm_t;
+    __m512i rand_index = _mm512_set_epi64(left + size,
+                                          left + 2 * size,
+                                          left + 3 * size,
+                                          left + 4 * size,
+                                          left + 5 * size,
+                                          left + 6 * size,
+                                          left + 7 * size,
+                                          left + 8 * size);
+    zmm_t key_vec = vtype::template i64gather<sizeof(type_t)>(rand_index, keys);
+
+    index_type index_vec;
+    zmm_t sort;
+    index_vec = zmm_vector<uint64_t>::template i64gather<sizeof(uint64_t)>(
+            rand_index, indexes);
+    sort = sort_zmm_64bit<vtype>(key_vec, index_vec);
```

Comment:
pivot is picked just based on keys and not index, right? `index_vec` is never really used anywhere in this function, which means you do not need a new `get_pivot_64bit` function. You can move this function to `avx512-64bit-common.h` and just call that instead. You will also need to move `sort_zmm_64_bit<vtype>` into the common file. This should eliminate the need to calling gather function twice and save you some cpu cycles.